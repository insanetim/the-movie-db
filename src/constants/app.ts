export const APP_NAME = 'THE MOVIE DB'
export const API_URL = 'https://api.themoviedb.org/3/'
export const ACCESS_TOKEN_AUTH =
  'eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI1N2RiYjI4M2M4MDBjY2QwYmM5YTczNTJjN2Y1ZmEzYiIsInN1YiI6IjYzMmM4YThkZTNmYTJmMDA3ZmQ0ZjY5OSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.nMVY53gL3_w8JAHI-N_uKQyXpvT9TsZ4_p8nhJSnAdQ'
export enum NOTIFICATION_TYPE {
  ERROR = 'error',
  SUCCESS = 'success'
}
export const NOTIFICATION_DURATION = 2.5
