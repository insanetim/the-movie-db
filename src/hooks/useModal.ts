import { nanoid } from '@reduxjs/toolkit'
import { useCallback } from 'react'
import { useModalsContext } from 'src/contexts/ModalsProvider'
import { Modal } from 'src/contexts/ModalsProvider/types'

const useModal = (initModalId?: string) => {
  const { dispatch } = useModalsContext()

  const modalId = initModalId ?? nanoid()

  const openModal = useCallback(
    (modal: Omit<Modal, 'modalId'>) => {
      dispatch({ payload: { ...modal, modalId }, type: 'OPEN_MODAL' })
    },
    [dispatch, modalId]
  )

  const closeModal = useCallback(() => {
    dispatch({ payload: modalId, type: 'CLOSE_MODAL' })
  }, [dispatch, modalId])

  return { closeModal, openModal }
}

export default useModal
