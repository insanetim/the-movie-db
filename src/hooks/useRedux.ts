import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'

import type { AppDispatch, RootState } from '../store'

type DispatchFunc = () => AppDispatch
const useAppDispatch: DispatchFunc = useDispatch
const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export { useAppDispatch, useAppSelector }
