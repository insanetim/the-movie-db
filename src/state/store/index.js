import { configureStore } from '@reduxjs/toolkit'
import { createLogicMiddleware } from 'redux-logic'

import httpClient from 'src/api/httpClient'
import rootReducer from './reducer'
import appOperations from '../app/operations'
import sessionOperations from '../session/operations'
import dashboardOperations from '../dashboard/operations'
import listsOperations from '../lists/operations'
import watchlistOperations from '../watchlist/operations'
import favoritesOperations from '../favorites/operations'
import movieOperations from '../movie/operations'

const operations = [
  ...appOperations,
  ...sessionOperations,
  ...dashboardOperations,
  ...listsOperations,
  ...watchlistOperations,
  ...favoritesOperations,
  ...movieOperations
]
const operationsDependencies = { httpClient }
const logicMiddleware = createLogicMiddleware(operations, operationsDependencies)

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: null
    }).concat(logicMiddleware),
  devTools: process.env.NODE_ENV !== 'production'
})

export default store
