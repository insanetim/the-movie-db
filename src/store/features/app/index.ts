export * from './appSlice'
export * from './types'
