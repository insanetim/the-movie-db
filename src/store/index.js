import { configureStore } from '@reduxjs/toolkit'
import { createLogicMiddleware } from 'redux-logic'

import httpClient from 'src/api/httpClient'
import appReducer from './app'
import sessionReducer from './session'
import dashboardReducer from './dashboard'
import listsReducer from './lists'
import watchlistReducer from './watchlist'
import favoritesReducer from './favorites'
import movieReducer from './movie'
import appOperations from './app/operations'
import sessionOperations from './session/operations'
import dashboardOperations from './dashboard/operations'
import listsOperations from './lists/operations'
import watchlistOperations from './watchlist/operations'
import favoritesOperations from './favorites/operations'
import movieOperations from './movie/operations'

const operations = [
  ...appOperations,
  ...sessionOperations,
  ...dashboardOperations,
  ...listsOperations,
  ...watchlistOperations,
  ...favoritesOperations,
  ...movieOperations
]
const operationsDependencies = { httpClient }
const logicMiddleware = createLogicMiddleware(operations, operationsDependencies)

const store = configureStore({
  reducer: {
    app: appReducer,
    session: sessionReducer,
    dashboard: dashboardReducer,
    lists: listsReducer,
    watchlist: watchlistReducer,
    favorites: favoritesReducer,
    movie: movieReducer
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(logicMiddleware),
  devTools: process.env.NODE_ENV !== 'production'
})

export default store
