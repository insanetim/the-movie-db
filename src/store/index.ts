import { combineReducers, configureStore } from '@reduxjs/toolkit'
import {
  FLUSH,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  REHYDRATE,
  persistReducer,
  persistStore,
} from 'redux-persist'
import storage from 'redux-persist/lib/storage'

import appReducer from './app'
import { showModal } from './app/actions'
import authReducer from './auth'
import createdListsReducer from './createdLists'
import dashboardReducer from './dashboard'
import favoriteReducer from './favorite'
import listDetailsReducer from './listDetails'
import movieDetailsReducer from './movieDetails'
import watchlistReducer from './watchlist'

const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['account'],
}

const rootReducer = combineReducers({
  app: appReducer,
  auth: persistReducer(authPersistConfig, authReducer),
  createdLists: createdListsReducer,
  dashboard: dashboardReducer,
  favorite: favoriteReducer,
  listDetails: listDetailsReducer,
  movieDetails: movieDetailsReducer,
  watchlist: watchlistReducer,
})

export const store = configureStore({
  devTools: process.env.NODE_ENV !== 'production',
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          REHYDRATE,
          showModal.type,
        ],
      },
    }),
  reducer: rootReducer,
})

export const persistor = persistStore(store)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
