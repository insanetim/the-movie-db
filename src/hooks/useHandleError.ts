import { NOTIFICATION_TYPE } from 'src/constants/app'
import { showNotification } from 'src/store/features/app'
import { useAppDispatch } from 'src/store/hooks'
import errorMessage from 'src/utils/helpers/errorMessage'

const useHandleError = () => {
  const dispatch = useAppDispatch()

  const handleError = (error: unknown) => {
    dispatch(
      showNotification({
        message: errorMessage(error),
        type: NOTIFICATION_TYPE.ERROR,
      })
    )
  }

  return { handleError }
}

export default useHandleError
