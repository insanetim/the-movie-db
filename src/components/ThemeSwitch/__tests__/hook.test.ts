import { act, renderHook } from '@testing-library/react'
import { selectTheme, setTheme, Theme } from 'src/store/features/app'
import { useAppDispatch, useAppSelector } from 'src/store/hooks'

import useContainer from '../hook'

jest.mock('src/store/features/app')
jest.mock('src/store/hooks')

const mockUseAppDispatch = useAppDispatch as jest.MockedFunction<
  typeof useAppDispatch
>
const mockUseAppSelector = useAppSelector as jest.MockedFunction<
  typeof useAppSelector
>
const mockSelectTheme = selectTheme as jest.MockedFunction<typeof selectTheme>
const mockSetTheme = setTheme as jest.MockedFunction<typeof setTheme>

describe('ThemeSwitch useContainer hook', () => {
  const mockDispatch = jest.fn()

  beforeEach(() => {
    mockUseAppDispatch.mockReturnValue(mockDispatch)
    mockUseAppSelector.mockReturnValue('light')
    mockSelectTheme.mockReturnValue('light')
    mockSetTheme.mockImplementation(value => ({
      payload: value,
      type: 'app/setTheme',
    }))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('Redux integration', () => {
    it('should call useAppSelector with selectTheme', () => {
      renderHook(() => useContainer())

      expect(mockUseAppSelector).toHaveBeenCalledWith(mockSelectTheme)
    })
  })

  describe('Return values', () => {
    it('should expose currentTheme from selector', () => {
      const { result } = renderHook(() => useContainer())

      expect(result.current.currentTheme).toBe('light')
    })

    it('should reflect updated theme on rerender', () => {
      mockUseAppSelector
        .mockReturnValueOnce('light')
        .mockReturnValueOnce('dark')

      const { rerender, result } = renderHook(() => useContainer())

      expect(result.current.currentTheme).toBe('light')

      rerender()

      expect(result.current.currentTheme).toBe('dark')
    })
  })

  describe('handleChange', () => {
    it('should dispatch setTheme with provided value', () => {
      const { result } = renderHook(() => useContainer())
      const nextTheme: Theme = 'dark'

      act(() => {
        result.current.handleChange(nextTheme)
      })

      expect(mockDispatch).toHaveBeenCalledTimes(1)
      expect(mockDispatch).toHaveBeenCalledWith(mockSetTheme(nextTheme))
    })
  })
})
