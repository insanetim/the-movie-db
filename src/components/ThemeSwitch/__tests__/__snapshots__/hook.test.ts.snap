// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitch useContainer hook should match snapshot 1`] = `
{
  "currentTheme": "light",
  "handleChange": [Function],
}
`;
